src/Config/Parser/PreParserInterface.php

```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Overblog\GraphQLBundle\Config\Parser;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+interface PreParserInterface extends ParserInterface
+{
+    /**
+     * @param \SplFileInfo     $file
+     * @param ContainerBuilder $container
+     * @param array            $configs
+     *
+     * @return mixed
+     */
+    public static function preParse(\SplFileInfo $file, ContainerBuilder $container, array $configs = []);
```

Comment:
@Vincz I'm preparing the annotation branch to merge it in master but I don't understand why we need this method?

Reply:
Hi! It's for parsers that need to parse files in two times. With the annotations and the doctrine entities converted to types, we first need to have a map of entity => GQL types to handle associations and validation.

Reply:
Thank you, I will give it a better look to see if we can cache some more data on pre parse to speedup parse process.

Reply:
Hey @mcg-web ! Finally, I'm not very happy with the annotations and the parsing process. 

The preParse allowed me to have the mapping between PHP classes and the GraphQL types in order to auto-guess graphql type from doctrine associations or PHP type hints. So we generate Graphql types / queries from PHP classes or methods. 

So something like that:
```
/**
 * @GQL\Query
 */
public function getItem(int $id) : Item {
    return $this->find($id);
}
```
is automatically turn into a field config like
```
[
    'getItem' => [
        'type' => 'Item',
        'args' => ['id' => ['type' => 'Int!']],
        'resolve' => '@=service(...).getItem(args["id"])'
    ]
]
```
It is possible, because we know that the PHP class `Item` match the Graphql type `Item`.

With the annotations, I also introduced 'Query and Mutation' providers (see [Providers / Query / Mutation](https://github.com/overblog/GraphQLBundle/blob/annotations/docs/annotations/annotations-reference.md#provider)). It's quite useful as it allows to define fields for the root query or mutation from services (as long as the service are accessible in the container through their class name) and having the args and type auto-guessed is really nice too.
To handle this, it's done in the AnnotationParser. When the root query or the root mutation is detected, their fields are also populated from the providers. It works great, but it only works with annotations. If we define the Root Query or Mutation with an other format with a different parser (yaml or xml), it won't work. 

The last feature is the argument transformer (see [Arguments Transformer](https://github.com/overblog/GraphQLBundle/blob/annotations/docs/annotations/arguments-transformer.md)). It also uses the mapping PHP / GQL to be able to transform  Graphql Input into PHP arguments. It's nice because it populates automatically the method arguments and validate them.

At the moment, all of this features are very specific to the annotation parser.

It sucks. 

Maybe we should move all this feature at a higher level. The parsers (all types) would not only return config types but also class mapping and providers, like that for example:

```
Human:
    type: object
    class: App\Entity\Human
```

And it would also be able to return Query or Mutation providers (the properties not defined would be auto-guessed) like that for example:

```
$providers:
    App\Repository\ItemRepository:
        findAll:
            parentType: Query
        findOne:
            parentType: Query
            type: Item
            args: { id: Int! }
            resolve: "@=service("App\Repository\ItemRepository").findOne(args['id'])"
```

So, the parsing process, would become:
1. Parse to get the mapping between PHP classes & GraphQL types.
2. Parse the fields configuration (and allow merge if a type is define in multiple places ?).
3. Parse the providers and add corresponding fields to the parent types.

This way, it will not be specific to annotation anymore.  

Tell me what you think about this !

Reply:
Ok, so for now, I'll just add the fields builders on the annotations and we will be good to go.  We'll check the rest later.